---
title: "Downstream Analysis"
author: "DIVYA SHARMA"
date: "2026-02-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
#Load the packages

library(Seurat)
library(SeuratDisk)
library(dplyr)
library(R.utils)  
library(ggplot2)
library(ggExtra)
library(RColorBrewer)
library(openxlsx)
library(dplyr)
library(scales)
library(HGNChelper)
library(dittoSeq)
library(harmony)
library(batchelor)
library(zellkonverter)
library(SingleCellExperiment)
```

```
```{r}
#Define Directories

inputDir <- "~/GSE279086/input"
outputDir <- "~/GSE279086/output"
plotDir <- "~/GSE279086/plots"
inputDir
```

```{r}

#Merge all Seurat objects into one}
length(seurat_list)
names(seurat_list)
print(names(seurat_list))
seurat_combined <- merge(seurat_list[[1]], y = seurat_list[-1], add.cell.ids = names(seurat_list))
unique(seurat_combined$orig.ident)

saveRDS(seurat_combined, file =  paste0(outputDir,"01_GSE279086_seurat_combined.rds"))
seurat_combined <- readRDS(file = paste0(outputDir, "01_GSE279086_seurat_combined.rds"))

# Explore Combined Seurat Object}
class(seurat_combined)            # 'Seurat'
Assays(seurat_combined)           # simpleAssays
DefaultAssay(seurat_combined)     # RNA
dim(seurat_combined)              # 28317 genes/features x 181842 cells
```

```{r}
# Adding 'Condition' col to the seurat object to do qc plot much better
table(seurat_combined$orig.ident)
gsm_ids <- unique(seurat_combined$orig.ident)

condition_map <- c(
  rep("AKI", 12),
  rep("DKD", 14),
  rep("HCKD", 3),
  rep("Healthy", 20)
)

names(condition_map) <- gsm_ids

# Create cell-level condition vector (drop names!)
seurat_combined$Condition <- unname(condition_map[seurat_combined$orig.ident])

table(seurat_combined$Condition) 
# AKI     DKD    HCKD Healthy 
#58517   85982    8541   28802 

sum(is.na(seurat_combined$Condition)) 
#0
```

```{r}
head(rownames(seurat_combined), 20)
seurat_combined[["percent.mt"]] <- PercentageFeatureSet(
  seurat_combined,
  pattern = "^MT-"
)

seurat_combined[["percent.rb"]] <- PercentageFeatureSet(
  seurat_combined,
  pattern = "^RPL|^RPS"
)

colnames(seurat_combined@meta.data)

summary(seurat_combined$percent.mt)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.299  25.243  39.858  42.017  56.832  98.661

summary(seurat_combined$percent.rb)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.000   3.224   5.461   6.603   8.722  51.393
```

```{r}
#.................Perform QC..............Pre QC graphs#
# These parameters define quality control rules that help distinguish real, healthy single cells from empty droplets, dying cells, or technical artifacts.
```{r}
study_id <- "GSE279086"
seurat_combined <- AddMetaData(seurat_combined, metadata = seurat_combined$orig.ident, col.name = "Sample")
Y
features <- c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb")

for (feat in features) {
  if (!feat %in% colnames(meta)) {
    warning(paste("Skipping", feat, "- not found in metadata"))
    next
  }
  print(summary(meta[[feat]]))
}


save_violin_plots_separate <- function(
    seurat_obj,
    plotDir,
    study_id,
    features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.rb")
) 
  
  if (!dir.exists(plotDir)) {
    dir.create(plotDir, recursive = TRUE)
  }

meta <- seurat_combined@meta.data
meta$sample <- as.factor(meta$orig.ident)

for (feat in features) {
  
  if (!feat %in% colnames(meta)) {
    warning(paste("Skipping", feat, "- not found in metadata"))
    next
  }}

p <- ggplot(meta, aes(x = sample, y = .data[[feat]])) +
  geom_violin(trim = TRUE, fill = "steelblue", alpha = 0.7) +
  geom_boxplot(width = 0.1, outlier.shape = NA, alpha = 0.6) +
  labs(
    title = feat,
    x = "Sample",
    y = feat
  ) +
  theme_bw(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5)
  )

ggsave(
  filename = file.path(
    plotDir,
    paste0(study_id, "_preQC_", feat, "_violin.png")
  ),
  plot = p,
  width = 16,
  height = 9,
  dpi = 400,
  bg = "white"
)


save_violin_plots_separate(seurat_combined, plotDir, study_id)

# Scatter plot of nCount vs nFeature with marginal histograms
density_scatter_plot <- function(seurat_obj, filename){
  df <- data.frame(
    log1p_nCount_RNA = log1p(seurat_obj$nCount_RNA),
    log1p_nFeature_RNA = log1p(seurat_obj$nFeature_RNA),
    Condition = seurat_obj$Condition
  )
  
  p <- ggplot(df, aes(x = log1p_nCount_RNA, y = log1p_nFeature_RNA, colour = Condition)) +
    geom_point(alpha = 0.3, size = 0.5) +
    theme_minimal() +
    theme(legend.position.inside = c(0.05, 0.95), legend.justification = c("left", "top"), legend.key.size = unit(0.5, 'cm')) +
    guides(color = guide_legend(override.aes = list(size = 5))) + # 'size' here controls the symbol size
    labs(x = "log1p(nCount_RNA)", y = "log1p(nFeature_RNA)")
  
  p <- ggMarginal(p, type = "histogram", fill = "skyblue", bins = 40)
  ggsave(
    filename,
    plot = p,
    width = 8,
    height = 10,
    dpi = 400,
    bg = "white"
  )
  
}

density_scatter_plot(seurat_combined, file.path(plotDir, paste0(study_id, "_preQC_density-scatter.png")))

```
```{r}
# Remove cells with abnormal total RNA counts; automatically detects outliers
fil_nCounts <- function(seurat_combined, threshold_mahalanobis){
  ncounts_data <- as.data.frame(seurat_obj@meta.data$nCount_RNA)
  colnames(ncounts_data) <- "nCount_RNA"
  
  mahalanobis_dist <- mahalanobis(
    x = ncounts_data,
    center = colMeans(ncounts_data),
    cov = cov(ncounts_data)
  )
  
  seurat_obj$mahal_dist_nCount <- mahalanobis_dist
  mahal.fil <- qchisq(threshold_mahalanobis, df = ncol(ncounts_data))
  message(paste0("Mahalanobis threshold: ", round(mahal.fil, 3)))
  
  cells_to_remove <- rownames(seurat_obj@meta.data)[seurat_obj$mahal_dist_nCount > mahal.fil]
  message(paste0("Removing ", length(cells_to_remove), " cells (outliers by nCount_RNA)"))
  
  return(subset(seurat_obj, cells = setdiff(colnames(seurat_obj), cells_to_remove)))
}

# Filter low (empty droplets or poor capture) & high (likely doublets) gene cells
filter_nFeatures <- function(seurat_obj, min_feat, max_feat, ...) {
  return(subset(seurat_obj, subset = nFeature_RNA > min_feat & nFeature_RNA < max_feat))
}

# Filter high mitochondrial RNA 
filter_mt <- function(seurat_obj, max_mt) {
  return(subset(seurat_obj, subset = percent.mt < max_mt))
}

# Filter high ribosomal RNA 
filter_rb <- function(seurat_obj, max_rb) {
  return(subset(seurat_obj, subset = percent.rb < max_rb))
}

```
```{r QC thresholds}
#Putting Thresholds...
study_id <- "GSE279086"
min_features <- 200 # A cell must express at least 200 genes to be considered real. <200 --> empty droplets/cause most noise
max_features <- 7000 # Cells with more than 7000 genes are removed. Extremely high gene counts often indicate doublets
max_percent_mt <- 15 # Remove cells where >20% of RNA comes from mitochondria
max_percent_rb <- 10 # Remove cells dominated by ribosomal expression. High rRNA --> low info content/technical bias
threshold_mahalanobis <- 0.95 # Remove the worst 5% of cells that look abnormal overall (catches subtle outliers)

seurat_filtered <- filter_nFeatures(seurat_combined, min_features, max_features)

seurat_filtered <- filter_mt(seurat_filtered, max_percent_mt)

seurat_filtered <- fil_nCounts(seurat_filtered, threshold_mahalanobis)

seurat_filtered <- filter_rb(seurat_filtered,max_percent_rb)

# Cell counts before QC
df_pre_qc <- as.data.frame(table(seurat_combined$Sample))
colnames(df_pre_qc) <- c("Sample_ID", "Cells_Before_QC")

# Cell counts after QC
df_post_qc <- as.data.frame(table(seurat_filtered$Sample))
colnames(df_post_qc) <- c("Sample_ID", "Cells_After_QC")

# Merge into final QC table (left join behavior)
final_qc_table <- merge(
  df_pre_qc,
  df_post_qc,
  by = "Sample_ID",
  all.x = TRUE
)

final_qc_table
total_cells_before_qc <- sum(final_qc_table$Cells_Before_QC, na.rm = TRUE)
total_cells_after_qc <- sum(final_qc_table$Cells_After_QC, na.rm = TRUE)

total_cells_before_qc #181842
total_cells_after_qc
```

```{r}
#Save Summary Table}

# create folder if not exists
if (!dir.exists(outputDir)) {
  dir.create(outputDir, recursive = TRUE)
}

qc_metrics_file <- file.path(outputDir, paste0(study_id, "_QC_metrics.csv"))
write.csv(final_qc_table, qc_metrics_file, row.names = FALSE)
cat("Saved QC summary table to:", qc_metrics_file, "\n")
```

```{r}
#...........................Post QC graphs...........................................
study_id <- "GSE279086"

save_violin_plots_separate <- function(
    seurat_obj,
    plotDir,
    study_id,
    features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.rb")
) {
  
  # Ensure output directory exists
  if (!dir.exists(plotDir)) {
    dir.create(plotDir, recursive = TRUE)
  }
  
  meta <- seurat_obj@meta.data
  meta$sample <- as.factor(meta$orig.ident)
  
  for (feat in features) {
    
    if (!feat %in% colnames(meta)) {
      warning(paste("Skipping", feat, "- not found in metadata"))
      next
    }
    
    p <- ggplot(meta, aes(x = sample, y = .data[[feat]])) +
      geom_violin(trim = TRUE, fill = "steelblue", alpha = 0.7) +
      geom_boxplot(width = 0.1, outlier.shape = NA, alpha = 0.6) +
      labs(
        title = feat,
        x = "Sample",
        y = feat
      ) +
      theme_bw(base_size = 14) +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5)
      )
    
    ggsave(
      filename = file.path(
        plotDir,
        paste0(study_id, "_postQC_", feat, "_violin.png")
      ),
      plot = p,
      width = 16,
      height = 9,
      dpi = 400,
      bg = "white"
    )
  }
}

save_violin_plots_separate(seurat_filtered, plotDir, study_id)

# Scatter plot of nCount vs nFeature with marginal histograms
density_scatter_plot <- function(seurat_obj, filename){
  df <- data.frame(
    log1p_nCount_RNA = log1p(seurat_obj$nCount_RNA),
    log1p_nFeature_RNA = log1p(seurat_obj$nFeature_RNA),
    Condition = seurat_obj$Condition
  )
  
  p <- ggplot(df, aes(x = log1p_nCount_RNA, y = log1p_nFeature_RNA, colour = Condition)) +
    geom_point(alpha = 0.3, size = 0.5) +
    theme_minimal() +
    theme(legend.position.inside = c(0.05, 0.95), legend.justification = c("left", "top"), legend.key.size = unit(0.5, 'cm')) +
    guides(color = guide_legend(override.aes = list(size = 5))) + # 'size' here controls the symbol size
    labs(x = "log1p(nCount_RNA)", y = "log1p(nFeature_RNA)")
  
  p <- ggMarginal(p, type = "histogram", fill = "skyblue", bins = 40)
  ggsave(
    filename,
    plot = p,
    width = 8,
    height = 10,
    dpi = 400,
    bg = "white"
  )
  
}

density_scatter_plot(seurat_filtered, file.path(plotDir, paste0(study_id, "_postQC_density-scatter.png")))
```

```{r}
#This step prioritizes functionally relevant transcripts and removes low-information pseudogenes and non-coding noise, thereby improving the signal-to-noise ratio in dimensionality reduction
protein_coding_genes <- readLines("C:/Users/loneb/OneDrive/Documents/GSE279086/input/Protein_coding_genes.txt")

# Subset BEFORE normalization
seurat_filtered <- subset(seurat_filtered, features = protein_coding_genes)

####  Explore Filtered Seurat Object after QC Filter  ####
class(seurat_filtered)            # Seurat class
Assays(seurat_filtered)           # assays available in the rds
DefaultAssay(seurat_filtered)     # default/active assay in the rds
dim(seurat_filtered)              # 16442 genes/features x 5405 cells

saveRDS(seurat_filtered, file = paste0(outputDir,"02_GSE279086_seurat_qc_filtered.rds"))
```

```{r Load again the Filtered Seurat Object}
seurat_filtered <- readRDS(file = paste0(outputDir, "02_GSE279086_seurat_qc_filtered.rds"))
```


```{r Standard workflow for normalization and scaling}
#.................Normalization.........................
cat("Normalizing data using LogNormalize method (scale factor = 10,000)...\n")
seurat_processed <- NormalizeData(seurat_filtered,
                                  normalization.method = "LogNormalize")  # What percentage of total RNA does each gene contribute in this cell?

seurat_processed <- FindVariableFeatures(seurat_processed, selection.method = "vst", nfeatures = 2500)
seurat_processed <- ScaleData(seurat_processed) 
seurat_processed <- RunPCA(seurat_processed, dims=1:100, npcs = 100)
```

```{r Custom colors for Condition in metadata}
custom_colors <- c("AKI"="purple3", "DKD"="darkorange3", "HCKD"="green4", "Healthy"="blue3")

```{r Elbow Plot}
# Get the standard deviation for each PC then get the variance explained by each PC, then calculate the cumulative variance
stdev <- seurat_processed[["pca"]]@stdev
var_explained <- stdev^2 / sum(stdev^2)
cum_var <- cumsum(var_explained)
pca_var_df <- data.frame(
  PC = 1:length(stdev),
  Variance = var_explained,
  CumulativeVariance = cum_var
)
pca_var_df
num_PCs <- min(which(cum_var >= 0.95)) 
num_PCs #83

elbow_plot <- ElbowPlot(seurat_processed, ndims = 100, reduction = 'pca')+
  labs(title = "PCA Elbow Plot for Dimensionality Selection")

ggsave(file.path(plotDir, paste0(study_id, "_ElbowPlot.png")),
       elbow_plot, width = 8, height = 6, bg = 'white')
```

```{r}
```{r Total variance explained across the top 35 PCs}
pca_stdev <- seurat_processed[["pca"]]@stdev
pca_var_explained <- (pca_stdev^2) / sum(pca_stdev^2) * 100
total_var<- sum(pca_var_explained[1:35])
cat("Total variance explained by top PCs:", round(total_var, 2), "%\n") #Total variance explained by top PCs: 78.49 %

library(glue)
seurat_processed <- RunUMAP(seurat_processed, dims = 1:35)
seurat_processed <- FindNeighbors(seurat_processed, dims = 1:35, reduction = "pca", graph.name = "pca_nn")

seurat_processed <- FindClusters(seurat_processed, resolution = 0.8,graph.name = "pca_nn", cluster.name = "pca_clusters") 
n_clusters <- length(unique(seurat_processed$pca_clusters))
cat(glue("Clustering complete. Number of clusters: {n_clusters}\n"))  # 24 clusters found

```

```{r}
head(seurat_processed@meta.data, 3)
```
```{r Save Unintegrated Seurat Object}
#This extracts the normalized counts and metadata from Seurat and "repackages" them into the SCE (single cell experiment) container. It ensures that your gene names, cell barcodes, and cluster IDs aren't lost during the move.
saveRDS(seurat_processed, file = paste0(outputDir,"03_GSE279086_seurat_pca_umap.rds"))
seurat_processed <- readRDS(file = paste0(outputDir, "03_GSE279086_seurat_pca_umap.rds"))

DefaultAssay(seurat_processed) <- "originalexp"
sce <- as.SingleCellExperiment(seurat_processed, assay = "originalexp")
dim(sce)
assayNames(sce)
reducedDimNames(sce)
head(colData(sce), 2)
```
```{r}
#By converting to SCE, you make the data "translatable" so that tools like zellkonverter can turn it into an .h5ad file (AnnData), which Python can read perfectly
h5seurat_name <- "03_GSE279086_seurat_pca_umap.h5seurat"
h5ad_name <- "03_GSE279086_seurat_pca_umap.h5ad"

library(SeuratDisk)
# Example patch for SaveH5Seurat
SaveH5Seurat_Fixed <- function(object, filename, ...) {
  # Create a temporary V4-style assay from the V5 counts layer
  object[["RNA_temp"]] <- CreateAssayObject(counts = object[["RNA"]]$counts)
  DefaultAssay(object) <- "RNA_temp"
  # Remove the V5 assay that causes the error
  object[["RNA"]] <- NULL 
  SeuratDisk::SaveH5Seurat(object, filename = filename, ...)
}

BiocManager::install("anndataR")
BiocManager::install("rhdf5")
library(rhdf5)
library(anndataR)
writeH5AD(sce, file = paste0(outputDir, "GSE279086_seurat_pca_umap.h5ad"))
```



```{r Run Harmony & Perform Clustering on Harmony integrated data}
#here we perform batch correction to remove technical noise caused by different processing times or sequencing runs, ensuring samples overlap correctly. This allows us to identify true biological differences (like cell types) rather than seeing clusters driven by experimental artifacts.
harmony_processed <- RunHarmony(seurat_processed, c("Sample"), plot_convergence = TRUE)

harmony_processed <- RunUMAP(harmony_processed, reduction = "harmony", dims = 1:50)

harmony_processed <- FindNeighbors(harmony_processed, reduction = "harmony", dims = 1:50, graph.name = "harmony_nn")

harmony_processed <- FindClusters(harmony_processed, graph.name = "harmony_nn", resolution = 0.8, group.name = "Harmony_clusters")
#found 19 clusters

saveRDS(harmony_processed, file = paste0(outputDir,"GSE279086_harmony_corrected.rds"))
harmony_processed <- readRDS(file = paste0(outputDir, "GSE279086_harmony_corrected.rds"))

# 1. Convert the old 'originalexp' to the new 'Assay5' class
# This is the "Higher Version" upgrade step
harmony_processed[["RNA"]] <- as(harmony_processed[["originalexp"]], "Assay5")

# 2. Set RNA as the active assay
DefaultAssay(harmony_processed) <- "RNA"

merged1 = JoinLayers(harmony_processed) 
sce_harmony <- as.SingleCellExperiment(merged1, assay = "RNA")
dim(sce_harmony) # 16442  5405
assayNames(sce_harmony)
reducedDimNames(sce_harmony)
head(colData(sce_harmony),2)

#By doing this, you are saving all your hard work including Harmony integration, the UMAP coordinates, and the 19 clusters—into a .h5ad file that Python can open instantly
h5seurat_name1 <- "GSE279086_harmony_corrected.h5seurat"
h5ad_name1 <- "GSE279086_harmony_corrected.h5ad"

# Convert your v5 Seurat object to SingleCellExperiment
sce_v5 <- as.SingleCellExperiment(harmony_processed, assay = "RNA")

# 2. Write to H5AD (This is the 'Higher Version' file format)
h5ad_path <- file.path(outputDir, "GSE279086_v5_final.h5ad")
writeH5AD(sce_v5, file = h5ad_path)

message("Higher version H5AD saved: ", h5ad_path)
writeH5AD(sce_harmony, file = paste0(outputDir,  "GSE279086_harmony_corrected.h5ad"))

```

```{r}
```{r Dimention reduction plot on PCA}
p <- DimPlot(seurat_processed, reduction = "umap", group.by = "Sample") + ggtitle("Uncorrected") +  NoLegend()
ggtitle("Uncorrected")
ggsave(filename = file.path(plotDir, "BENCHMARKING_GSE297086_RawPCA_sample.png"), width = 7, height = 6, dpi = 600)
p

p <-DimPlot(seurat_processed, reduction = "umap", group.by = "Condition") +
  ggtitle("Uncorrected")
ggsave(filename = file.path(plotDir, "BENCHMARKING_GSE297086_RawPCA_condition.png"), width = 8, height = 6, dpi = 600)
p
```
```{r}
#harmony_processed <- RunUMAP(harmony_processed, reduction = "harmony", dims = 1:30)
p2 <-DimPlot(harmony_processed, reduction = "umap", group.by = "Condition") + ggtitle("Harmony") 
ggsave(filename = file.path(plotDir, "BENCHMARKING_GSE279086_Harmony_condition.png"), width = 8, height = 6, dpi = 600)
p2

p <- DimPlot(harmony_processed, reduction = "umap", group.by = "Sample") + ggtitle("Harmony") +  NoLegend()
ggsave(filename = file.path(plotDir, "BENCHMARKING_GSE279086_Harmony_sample.png"), width = 7, height = 6, dpi = 600)
p

umap_hcoords <- Embeddings(harmony_processed, "umap")
write.csv(umap_hcoords, file="GSE279086_umap_coordinates.csv")
```

```{r} 
#Combine Bar Plot
library(ggplot2)

compute_knn_batch_mixing <- function(
    seu,
    batch_var,
    reduction = "pca",
    dims = 1:50,
    k = 20
){
  
  
  if (!batch_var %in% colnames(seu@meta.data)) {
    stop(paste0("Metadata column '", batch_var, "' not found in seu@meta.data"))
  }
  
  if (!reduction %in% Reductions(seu)) {
    message("[INFO] PCA not found. Running Normalize → HVG → Scale → PCA")
    seu <- NormalizeData(seu, verbose = FALSE)
    seu <- FindVariableFeatures(seu, selection.method = "vst", nfeatures = 3000)
    seu <- ScaleData(seu, verbose = FALSE)
    seu <- RunPCA(seu, npcs = max(dims), verbose = FALSE)
  }
  
  emb <- Embeddings(seu, reduction = reduction)[, dims, drop = FALSE]
  nn  <- FNN::get.knn(emb, k = k)$nn.index
  labs <- seu@meta.data[[batch_var]]
  
  same <- sapply(seq_len(nrow(nn)), function(i){
    mean(labs[nn[i,]] == labs[i], na.rm = TRUE)
  })
  
  df <- data.frame(batch = labs, frac_same_batch = same)
  aggregate(frac_same_batch ~ batch, df, mean)
}

plot_knn_batch_mixing <- function(mix_df){
  ggplot(mix_df, aes(x = batch, y = frac_same_batch)) +
    geom_col(fill = "steelblue") +
    labs(
      title = "Batch Mixing",
      x = "Batch",
      y = "Mean same-batch fraction (higher = stronger batch effect)"
    ) +
    theme_minimal(base_size = 12)
}


batch_column <- "Sample"

mix_df <- compute_knn_batch_mixing(
  seu       = seurat_processed,             
  batch_var = batch_column,
  reduction = "pca",
  dims      = 1:50,
  k         = 20
)

mix_df_harmony <- compute_knn_batch_mixing(
  seu       = harmony_processed,
  batch_var = batch_column,
  reduction = "harmony",
  dims      = 1:50,
  k         = 20
)


mix_df_combined <- rbind(mix_df, mix_df_harmony)


mix_df_combined1 <- rbind(
  transform(mix_df, Status = "Before"),
  transform(mix_df_harmony, Status = "Harmony")
)


```

```{r}
#This step creates a diagnostic bar chart to visually and quantitatively evaluate the success of your batch correction with Harmony. It is a standard quality control procedure in single-cell RNA-seq workflows to ensure that technical noise is reduced without erasing biological signals.
batch_cor_plot <- ggplot(mix_df_combined1, aes(x = batch, y = frac_same_batch, fill = Status)) +
  geom_col(position = position_dodge(width = 0.8), width = 0.7) +
  labs(
    title = "Batch Mixing",
    x = "Samples",
    y = "Mean same-batch fraction (higher = stronger batch effect)"
  ) +
  scale_fill_manual(
    values = c(
      "Before" = "#00CED1",     # turquoise
      "Harmony" = "#9370DB"     # new purple shade (example 3rd color)
    )
  ) +
  theme_minimal(base_size = 15) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.title = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

batch_cor_plot

ggsave(filename = paste0(plotDir,"batch_correction_barplot_combined.png"), plot = batch_cor_plot, width = 23, height = 8, dpi = 300)
```


